zookeeper-server/src/main/java/org/apache/zookeeper/Op.java

```diff
@@ -253,6 +254,142 @@ void validate() throws KeeperException {
         PathUtils.validatePath(path);
     }
 
+    /**
+     * "Fluent builder" for {@link Create} operations.
+     */
+    public static class CreateBuilder {
+        private String path;
+        private byte[] data;
+        private List<ACL> acl;
+        private Integer createModeFlag;
+        private CreateMode createMode;
+        private boolean returnStat = false;
+        private Long ttl;
+
+        /**
+         * Sets the path for the node.  Must be set before calling
+         * {@link #build()}.
+         *
+         * @param path  the path for the node
+         * @return the builder, for chaining.
+         */
+        public CreateBuilder setPath(String path) {
+            this.path = path;
+            return this;
+        }
+
+        /**
+         * Sets the initial data for the node, or {@code null}.
+         *
+         * @param data  the initial data for the node
+         * @return the builder, for chaining.
+         */
+        @SuppressFBWarnings({"EI_EXPOSE_REP", "EI_EXPOSE_REP2"})
+        public CreateBuilder setData(byte[] data) {
+            this.data = data;
+            return this;
+        }
+
+        /**
+         * Sets the acl for the node.  Must be set before calling
+         * {@link #build()}.
+         *
+         * @param acl  the acl for the node
+         * @return the builder, for chaining.
+         */
+        public CreateBuilder setACL(List<ACL> acl) {
+            this.acl = acl;
```

Comment:
are we allowing a `null` value here ?

Reply:
The idea was to perform validation at `build()` time—but it is true that `path` and `acl` are not, in fact, validated by the `Create`/`CreateTTL` constructors.  I suppose that should be fixed.

Or are you suggesting that a check should be added at *this* point?  (That would match your `int`/`long` comment, and correspond to a "properties must be set once, and must be set right" model—which is not what I was initially going for.)

Reply:
It is better to perform validation on build() when possible

Reply:
Done (checking `path` and `acl`).

I had added the checks to the `Op.Create` constructor, but went back to checking at `build()` time, not to change the existing `create(...)` methods—which still do *not* check for `null`—as it may be that some programs accidentally do that.  (No error is thrown if the request is never submitted.)